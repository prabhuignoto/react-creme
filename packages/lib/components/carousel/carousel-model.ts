import React from 'react';

export type CarouselProps = {
  autoPlay?: number;
  border?: boolean;
  children: React.ReactNode | React.ReactNode[];
  direction?: 'horizontal' | 'vertical';
  enableSwipe?: boolean;
  focusable?: boolean;
  height?: number;
  size?: 'sm' | 'md' | 'lg';
  transition?: string;
};

export type CarouselItemProps = {
  height: number;
  id: string;
  left?: string;
  top?: string;
  visible?: boolean;
  width: number;
};

export type CarouselItemsProps = Pick<
  CarouselProps,
  'children' | 'direction' | 'height' | 'size'
> & {
  activePage: number;
  carouselItems: CarouselItemProps[];
  totalItems: number;
  width: number;
};

export type CarouselTrackProps = Pick<
  CarouselProps,
  'direction' | 'focusable' | 'size'
> & {
  activeIndex: number;
  focusable?: boolean;
  handleSelection: (index: number) => void;
  hideNext?: boolean;
  hidePrevious?: boolean;
  length: number;
  onNext: () => void;
  onPrevious: () => void;
};

export type CarouselButtonProps = Pick<
  CarouselProps,
  'direction' | 'focusable' | 'size'
> & {
  hide?: boolean;
  label: string;
  onClick: () => void;
  position: 'left' | 'right';
};
