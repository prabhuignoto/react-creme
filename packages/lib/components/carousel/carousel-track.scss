@use '../../design/core.scss';
@use '../../design/list.scss';
@use '../../design/theme.scss';

.rc-carousel-track {
  &.rc-carousel-track-horizontal {
    @extend %center;
    @extend %list-horizontal;
  }

  &.rc-carousel-track-vertical {
    @extend %list;
    width: 3rem;
  }
}

.rc-carousel-track-item {
  border: 1px solid theme.$secondary;
  border-radius: 50%;
  height: 10px;
  margin: 4px;
  width: 10px;
}

.rc-carousel-track-item:hover {
  cursor: pointer;
}

.rc-carousel-track-item-selected {
  background: theme.$primary;
  border: 1px solid theme.$primary;
}

.rc-carousel-icon {
  @extend %icon-md;
}

.rc-carousel-track-wrapper-horizontal {
  @extend %center;
}

.rc-carousel-track-wrapper-vertical {
  @extend %list;
}
