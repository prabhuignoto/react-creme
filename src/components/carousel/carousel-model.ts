import React from "react";

export interface CarouselModel {
  children: React.ReactNode | React.ReactNode;
  direction?: "horizontal" | "vertical";
  height?: number;
}

export interface CarouselItemModel {
  height?: number;
  id: string;
  left?: string;
  top?: string;
  visible?: boolean;
  width?: number;
}

export interface CarouselItemsModel {
  activePage: number;
  carouselItems: CarouselItemModel[];
  children: React.ReactNode[] | React.ReactNode;
  direction: "horizontal" | "vertical";
  height: number;
  totalItems: number;
  width: number;
}
