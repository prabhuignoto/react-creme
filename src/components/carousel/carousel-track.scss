@use "../../design/layout.scss";
@use "../../design/list.scss";
@use "../../design/colors.scss";
@use "../../design/icon.scss";

.rc-carousel-track {
  &.rc-carousel-track-horizontal {
    @extend %center;
    @extend %list-horizontal;
  }

  &.rc-carousel-track-vertical {
    @extend %list;
    width: 3rem;
  }
}

.rc-carousel-track-item {
  @extend %shadow-inset;
  border: 1px solid colors.$secondary;
  border-radius: 50%;
  height: 10px;
  margin: 4px;
  width: 10px;
}

.rc-carousel-track-item:hover {
  cursor: pointer;
}

.rc-carousel-track-item-selected {
  background: colors.$primary;
  border: 1px solid colors.$primary;
}

.rc-carousel-icon {
  @extend %icon-sm;
}

.rc-carousel-track-wrapper-horizontal {
  @extend %center;
}

.rc-carousel-track-wrapper-vertical {
  @extend %list;
}
