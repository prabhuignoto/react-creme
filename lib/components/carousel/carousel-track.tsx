import classNames from "classnames";
import React, { useMemo } from "react";
import { CarouselButton } from "./carousel-button";
import { CarouselTrackModel } from "./carousel-model";
import "./carousel-track.scss";

const CarouselTrack: React.FunctionComponent<CarouselTrackModel> = ({
  length,
  handleSelection,
  activeIndex,
  direction,
  onNext,
  onPrevious,
  hidePrevious,
  hideNext,
  focusable,
}: CarouselTrackModel) => {
  const carouselTrackClass = useMemo(
    () => classNames(["rc-carousel-track", `rc-carousel-track-${direction}`]),
    []
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames("rc-carousel-track-wrapper", {
        [`rc-carousel-track-wrapper-${direction}`]: true,
      }),
    [direction]
  );

  const focusableProps = useMemo(() => {
    return {
      tabIndex: focusable ? 0 : -1,
      onKeyDown: (event: React.KeyboardEvent) => {
        if (event.key === "ArrowRight") {
          onNext();
        } else if (event.key === "ArrowLeft") {
          onPrevious();
        }
      },
    };
  }, []);

  return (
    <div className={carouselContainerClass}>
      <CarouselButton
        onClick={onPrevious}
        position="left"
        hide={hidePrevious}
        direction={direction}
        focusable={focusable}
      />
      <ul className={carouselTrackClass} role="list">
        {Array.from({ length }).map((_, index) => (
          <li
            key={index}
            role="listitem"
            {...focusableProps}
            className={classNames([
              "rc-carousel-track-item",
              index === activeIndex ? "rc-carousel-track-item-selected" : "",
            ])}
            onClick={() => handleSelection(index)}
          ></li>
        ))}
      </ul>
      <CarouselButton
        onClick={onNext}
        position="right"
        hide={hideNext}
        direction={direction}
        focusable={focusable}
      />
    </div>
  );
};

export { CarouselTrack };
