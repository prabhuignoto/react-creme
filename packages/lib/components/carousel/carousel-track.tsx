import classNames from 'classnames';
import React, { useMemo } from 'react';
import { CarouselButton } from './carousel-button';
import { CarouselTrackProps } from './carousel-model';
import styles from './carousel-track.module.scss';
import rootStyles from './carousel.module.scss';

const CarouselTrack: React.FunctionComponent<CarouselTrackProps> = ({
  length,
  handleSelection,
  activeIndex,
  direction,
  onNext,
  onPrevious,
  hidePrevious,
  hideNext,
  focusable,
}: CarouselTrackProps) => {
  const carouselTrackClass = useMemo(
    () =>
      classNames([
        styles.track,
        direction === 'horizontal'
          ? styles.track_horizontal
          : styles.track_vertical,
      ]),
    []
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames([
        rootStyles.track_wrapper,
        direction === 'horizontal'
          ? styles.track_wrapper_horizontal
          : styles.track_wrapper_vertical,
        // [styles[`track-wrapper-${direction}`]]: true,
      ]),
    [direction]
  );

  return (
    <div className={carouselContainerClass}>
      <CarouselButton
        onClick={onPrevious}
        position="left"
        hide={hidePrevious}
        direction={direction}
        focusable={focusable}
        label="Previous"
      />
      <ul className={carouselTrackClass} role="list">
        {Array.from({ length }).map((_, index) => (
          <li
            key={index}
            role="listitem"
            className={classNames([
              styles.track_item,
              index === activeIndex ? styles.track_item_selected : '',
            ])}
            onClick={() => handleSelection(index)}
          ></li>
        ))}
      </ul>
      <CarouselButton
        onClick={onNext}
        position="right"
        hide={hideNext}
        direction={direction}
        focusable={focusable}
        label="Next"
      />
    </div>
  );
};

export { CarouselTrack };
