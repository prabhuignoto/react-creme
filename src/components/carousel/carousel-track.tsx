import classNames from "classnames";
import React, { useMemo } from "react";
import { CarouselButton } from "./carousel-button";
import { CarouselTrackModel } from "./carousel-model";
import "./carousel-track.scss";

const CarouselTrack: React.FunctionComponent<CarouselTrackModel> = ({
  length,
  handleSelection,
  activeIndex,
  direction,
  onNext,
  onPrevious,
  hidePrevious,
  hideNext,
}: CarouselTrackModel) => {
  const carouselTrackClass = useMemo(
    () => classNames(["rc-carousel-track", `rc-carousel-track-${direction}`]),
    []
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames("rc-carousel-track-wrapper", {
        [`rc-carousel-track-wrapper-${direction}`]: true,
      }),
    [direction]
  );

  return (
    <div className={carouselContainerClass}>
      <CarouselButton
        onClick={onPrevious}
        position="left"
        hide={hidePrevious}
        direction={direction}
      />
      <ul className={carouselTrackClass} role="list">
        {Array.from({ length }).map((_, index) => (
          <li
            key={index}
            role="listitem"
            className={classNames([
              "rc-carousel-track-item",
              index === activeIndex ? "rc-carousel-track-item-selected" : "",
            ])}
            onClick={() => handleSelection(index)}
          ></li>
        ))}
      </ul>
      <CarouselButton
        onClick={onNext}
        position="right"
        hide={hideNext}
        direction={direction}
      />
    </div>
  );
};

export { CarouselTrack };
