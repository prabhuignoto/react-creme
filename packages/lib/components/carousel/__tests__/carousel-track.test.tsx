import { render, fireEvent } from '@testing-library/react';
import { CarouselTrack } from '../carousel-track';
import { CarouselTrackProps } from '../carousel-model';
import { vi } from 'vitest';

describe('CarouselTrack', () => {
  const defaultProps: CarouselTrackProps = {
    activeIndex: 0,
    direction: 'horizontal',
    focusable: true,
    handleSelection: vi.fn(),
    hideNext: false,
    hidePrevious: false,
    length: 5,
    onNext: vi.fn(),
    onPrevious: vi.fn(),
  };

  it('renders without crashing', () => {
    const { container } = render(<CarouselTrack {...defaultProps} />);
    expect(container.firstChild).toBeInTheDocument();
  });

  it('renders the correct number of track items', () => {
    const { getAllByRole } = render(<CarouselTrack {...defaultProps} />);
    const items = getAllByRole('listitem');
    expect(items.length).toBe(defaultProps.length);
  });

  it('calls the handleSelection function when a track item is clicked', () => {
    const { getAllByRole } = render(<CarouselTrack {...defaultProps} />);
    const items = getAllByRole('listitem');
    fireEvent.click(items[0]);
    expect(defaultProps.handleSelection).toHaveBeenCalledWith(0);
  });

  it('calls the onNext function when the next button is clicked', () => {
    const { getByLabelText } = render(<CarouselTrack {...defaultProps} />);
    const nextButton = getByLabelText('Next');
    fireEvent.click(nextButton);
    expect(defaultProps.onNext).toHaveBeenCalled();
  });

  it('calls the onPrevious function when the previous button is clicked', () => {
    const { getByLabelText } = render(<CarouselTrack {...defaultProps} />);
    const previousButton = getByLabelText('Previous');
    fireEvent.click(previousButton);
    expect(defaultProps.onPrevious).toHaveBeenCalled();
  });
});
