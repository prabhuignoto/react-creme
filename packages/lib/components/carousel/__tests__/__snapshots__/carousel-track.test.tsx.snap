// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Carousel Track > Should render the track snapshot 1`] = `
<ul
  class="_track_4fc414 _track_horizontal_4fc414"
  role="list"
>
  <li
    class="_track_item_4fc414"
    role="listitem"
  />
  <li
    class="_track_item_4fc414 _track_item_selected_4fc414"
    role="listitem"
  />
  <li
    class="_track_item_4fc414"
    role="listitem"
  />
</ul>
`;
